pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/bestpractices/JUnitAssertionsShouldIncludeMessageRule.java

```diff
@@ -65,8 +65,11 @@ public JUnitAssertionsShouldIncludeMessageRule() {
         checks.add(new AssertionCall("assertEquals", 3) {
             @Override
             protected boolean isException(ASTArguments node) {
-                ASTExpression firstArgument = node.getFirstDescendantOfType(ASTExpression.class);
-                return firstArgument.getType() == null || firstArgument.getType() == String.class;
+                List<ASTExpression> arguments = node.findDescendantsOfType(ASTExpression.class);
+                boolean isExceptionJunit4 = isStringTypeOrNull(arguments.get(0));
+                boolean isExceptionJunit5 = isStringTypeOrNull(arguments.get(2));
+
+                return isExceptionJunit4 || isExceptionJunit5;
```

Comment:
That's okay, but keep in mind, that it's actually not 100% correct:
We would need to check either for the first argument (junit4) or last argument (junit5) to be a string - after we determined the Junit version. But the combination "first argument is a String" and "JUnit5" would be wrong (and would not even compile).
However, I think, we can live with that simplification: We know here already, that we have in total 3 arguments. And the compiler will ensure, that the correct overload is used (where ever the String is).

Reply:
> However, I think, we can live with that simplification: We know here already, that we have in total 3 arguments. And the compiler will ensure, that the correct overload is used (where ever the String is).

That's what I think, PMD checks should be based on programs that can be compiled successfully.